students/Ramkumar_Rajanbabu/Lesson05/mailroom_p3.py

```diff
@@ -0,0 +1,143 @@
+                send_letters_all()
+            elif user_response == 4:
+                exit_program()
+            elif user_response not in range(1,4):
+                print("Please enter a valid number within 1-4!\n")
+            
+def send_thanks():
+    """Show a list of donors and add a new donor based on user response."""
+
+    name = str(input("Type in the full donor name (or 'list' to view a list of donor names): "))
+    if name == "list":
+        view_list()
+        send_thanks() #Reprompt to main menu
+    else:
+        add_donation(name)
+
+def create_report():
+    """Create a report (table with values)."""
+    
+    title_string = "{:<25s} | {:s} | {:s} | {:s}".format("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    title_bar = len(title_string) * "-"
+    
+    #List comprehension
+    summary = [(donor_name, sum(amount), len(amount), sum(amount) / len(amount)) for donor_name, amount in donors_db.items()]
+    summary = sorted(summary, key=sort_key, reverse=True)
+    
+    print()
+    print(title_string)
+    print(title_bar)
+    for donor_tuple in summary:
+        print("{:<25s}  ${:>11.2f}  {:>10d}  ${:>12.2f}".format(*donor_tuple)) #f_string
+
+def send_letters_all():
+    """Generates a thank you letter for each donor and writes each letter as a text file."""
+    
+    for donor_name, amount in donors_db.items():
+        file_name = donor_name + ".txt"
+        with open(file_name, 'w') as f:
+            f.write(write_letter(donor_name, sum(amount)))
+    print("\nA thank you letter to each donor is saved as a text file.")
+    
+def exit_program():
+    """Exit program."""
+    
+    print("\nYou chose to quit the program, good-bye!")
+    sys.exit() #Exit script
+             
+def view_list():
+    """To view donor names in a vertical list."""
+    
+    print("\nList of Donor Names:")
+    for donor_name in donors_db.keys():
+        print(donor_name) #Prints each name (key)
+    
+def write_letter(name, donation_amount):
+    """Generates letter based on donor name and donation amount.
+    
+    Args:
+        name: donor name
+        donation_amount (float): donation amount  
+    Returns:
+        letter (string): a letter in string format
+
+    """
+    
+    letter = """
+    Dear {:s},
+    
+    Thank you for choosing to donate to this group. A special thank you for your generous donation of ${:0.2f}.
+    
+    Sincerely, 
+    Donation Society""".format(name, donation_amount)
+    return letter
+    
+def add_donation(name):
+    """Add new donation amount to donor list based on user response.
+    
+    Args:
+        name: donor name
+    Returns:
+        None
+
+    """
+    
+    if name not in donors_db:
+        donors_db[name] = [] #donors_db[key] = [value]
+        
+    #Exception handling
+    try:
+        donation_amount = float(input("Type in the donation amount: $"))
+    except ValueError: #ValueError if not integer/float
+        print("Please enter an integer.")
```

Comment:
So we've got a bug here - though it prints out the error, if a user enters a string, what does the program do with that string?  If I then run the report again, I get kicked out.